druid/src/widget/aspect_ratio_box.rs

```diff
@@ -0,0 +1,279 @@
+    )]
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        if let Some(ref mut inner) = self.inner {
+            inner.event(ctx, event, data, env);
+        }
+    }
+
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, event, data, env)
+    )]
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &T, env: &Env) {
+        if let Some(ref mut inner) = self.inner {
+            inner.lifecycle(ctx, event, data, env)
+        }
+    }
+
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, old_data, data, env)
+    )]
+    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &T, data: &T, env: &Env) {
+        if let Some(ref mut inner) = self.inner {
+            inner.update(ctx, old_data, data, env);
+        }
+    }
+
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, bc, data, env)
+    )]
+    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints, data: &T, env: &Env) -> Size {
+        bc.debug_check("AspectRatioBox");
+
+        let bc = if bc.max() == bc.min() {
+            warn!("Box constraints are tight. Aspect ratio box will not be able to preserve aspect ratio.");
+
+            *bc
+        } else if bc.max().width == f64::INFINITY && bc.max().height == f64::INFINITY {
+            warn!("Box constraints are INFINITE. Aspect ratio box won't be able to choose a size because the constraints given by the parent widget are INFINITE.");
+
+            // should I do this or should I just choose a default size
+            // I size the box with the child's size if there is one
+            let size = match self.inner.as_mut() {
+                Some(inner) => inner.layout(ctx, &bc, data, env),
+                None => Size::new(500., 500.),
+            };
+
+            BoxConstraints::new(Size::new(0., 0.), size)
+        } else {
+            let (mut box_width, mut box_height) = (bc.max().width, bc.max().height);
+
+            if self.ratio < 1.0 {
+                if (box_height >= box_width && box_height * self.ratio <= box_width)
+                    || box_width > box_height
+                {
+                    box_width = box_height * self.ratio;
+                } else if box_height >= box_width && box_height * self.ratio > box_width {
+                    box_height = box_width / self.ratio;
+                } else {
+                    // I'm not sure if these sections are necessary or correct
+                    // dbg!("ratio can't be preserved {}", self.ratio);
+                    warn!("The aspect ratio cannot be preserved because one of dimensions is tight and the other dimension is too small: bc.max() = {}, bc.min() = {}", bc.max(), bc.min());
+                }
+
+                BoxConstraints::tight(Size::new(box_width, box_height))
+            } else if self.ratio > 1.0 {
```

Comment:
It's it an invariant that the ratio is <= 1.0?

Reply:
What is being asked here? Are you asking if it is an invariant that the ratio is <= 1.0?

Reply:
oops yes, sorry.

This comment was slightly out of date; I was confused because elsewhere in some comments it mentioned that `ratio` should be in the range 0..=1.0, but that wasn't being upheld in the implementation (and didn't make sense, so I assume it's just that the comments were out of date)

Reply:
Oh oops, I don't know what I was thinking when I wrote that comment.